chia/consensus/blockchain.py

```diff
@@ -769,12 +769,20 @@ async def validate_unfinished_block(
 
         return PreValidationResult(None, required_iters, conds, uint32(0))
 
-    def contains_block(self, header_hash: bytes32) -> bool:
+    def contains_block(self, header_hash: bytes32, height: Optional[uint32] = None) -> bool:
         """
         True if we have already added this block to the chain. This may return false for orphan blocks
```

Comment:
this comment should be updated to say that it *will* return false for orphan blocks. I take the current "may" to mean that orphan blocks are less likely to be in the cache.

This function also always returned false for blocks that aren't in the cache, regardless of whether they're in the main chain or not. The comment should say this too.

The new `height` parameter isn't really explained either. I assume it adds an extra check; not only that `header_hash` is in the main chain and in the cache, but also that it has the specified height.

Reply:
the hight is an optimization for most cases where we now the hight of the block (the only case we dont is unfinished blocks)

i agree with all your points, im still trying to figure out exactly what the correct Behavior is supposed to be but im pretty sure the current state is incorect for both short_sync_backtrack and short_sync_batch